EIPS/eip-6865.md

```diff
@@ -0,0 +1,151 @@
+
+To overcome these challenges, we propose a standardized, on-chain solution that can accommodate the diverse and ever-changing web3.0 ecosystem. This approach would enhance scalability, reliability, and maintainability by shifting the responsibility of visualizing EIP-712 payloads from the wallets to the protocols themselves. Consequently, wallets can use a consistent and effective approach to EIP-712 message visualization.
+
+The adoption of a universal solution will not only streamline the efforts and reduce the maintenance burden for wallet providers, but it will also allow for faster and more extensive coverage across the ecosystem. This will ultimately result in users gaining a clearer understanding of the transactions they're signing, leading to increased security and an improved overall user experience within the crypto space.
+
+currently, most of the wallets display something similar to image below
+
+![](../assets/eip-6865/current-EIP-712-signature-wallet-interface.png)
+
+with visualization we can achieve something similar to image below where more insightful details are revealed to user thanks to the structured data returned from the EIP
+
+![](../assets/eip-6865/vision-EIP-712-signature-wallet-interface.png)
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+Each new protocol that uses EIP-712 MUST include the `visualizeEIP712Message` function in the  `verifyingContract` implementation
+
+```solidity
+/**
+* @notice This function processes an EIP-712 payload message and returns a structured data format emphasizing the potential impact on users' assets.
+* @dev The function returns assetsOut (assets the user is offering), assetsIn (assets the user would receive), and liveness (validity duration of the EIP-712 message).
+* @param encodedMessage The ABI-encoded EIP-712 message (abi.encode(types, params)).
+* @param domainHash The hash of the EIP-712 domain separator as defined in the EIP-712 proposal; see https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator.
+* @return Result struct containing the user's assets impact and message liveness.
+*/
+function visualizeEIP712Message(
+    bytes memory encodedMessage,
+    bytes32 domainHash
+) external view returns (Result memory);
+```
+
+### Params
+
+`encodedMessage` is bytes that represents the encoded EIP-712  message with `abi.encode` and it can be decoded using `abi.decode`
+
+`domainHash` is the bytes32 hash of the EIP-712 domain separator as defined in the EIP-712 proposal
+
+### Outputs
+
+The function MUST returns Result ; a struct that contains information's about user’s assets impact and the liveness of such a message if get signed.
```

Comment:
```suggestion
The function MUST return `Result`, a struct that contains information's about user’s assets impact and the liveness of such a message if it gets signed.
```

Reply:
fixed in commit [92b1ab4fb](https://github.com/ethereum/EIPs/pull/6865/commits/92b1ab4fb061db98e5a2f166034dba633e6ee297)